# Data Consistency Checks

## - Mixed-type Data
## - Missing Values
## - Duplicates

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# import data: products.csv from Original Data
path = r'/Users/azadehrezaei/Desktop/Career Foundry/Achievement 4/08-2021 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

#import data: orders_wrangled.csv from Prepared data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
# Change data type to string 
df_test['mix'] = df_test['mix'].astype('str')

# Finding Missing Values:

In [9]:
# Fidning the columns that contain missing values, and how many values they're missing:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# Creating a subset for the dataframe, containing the missing records:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Finding duplicates and assigning them to a new subset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
# Dropping duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exercise 4.5:

In [19]:
# Question 1: All consistency checks are done in the above section. 

In [20]:
# Question 2: 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##I don't think anything looks off in this dataframe in the first 3 columns based on the results of df.describe(), the only think I'd probably change is the data type of product_id, aisle_id and department_id columns since these columns are simply identifiers and although the values are integers they don't need to be treated as integers and there's no value in performing descriptive analysis on them. I would change these datatypes to string instead. 

The price column however shows inconsistency. The maximum price value is recorded as 9999 although the mean value is still pretty low. This means that this value might be an outlier.

In [21]:
# Question 3: Checking for mixed-type data in df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [22]:
# Question 3: Checking for mixed type data in df_ords:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_ords[weird]) > 0:
        print(col + ' has mixed data type')
    else:
        print(col + ' has one data type!')
    

Unnamed: 0 has one data type!
order_id has one data type!
user_id has one data type!
order_number has one data type!
order_day_of_week has one data type!
order_hour_of_day has one data type!
days_since_prior_order has one data type!


df_ords columns all have uniform data types. 

In [23]:
# Question 5: Checking for missing values in df_ords: 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The column "days_since_prior_order" has 206,209 rows of missing values. 
This number can be the first time users placed their order which means since they are the very first order, this row doesn't get any value!


In [24]:
df_ords.shape

(3421083, 7)

To check this hypothesis we can compare the number of missing values in this column with the number of order_id's that are marked as 1:

In [25]:
df_ords_firstorders = df_ords[df_ords['order_number'] == 1]

In [26]:
df_ords_firstorders.shape

(206209, 7)

The number of missing values in the "days_since_prior_order" column and the number of first orders are both 206,209. So this shows that the empty/missing values in "days_since_prior_order" column are simply the users' first orders. 
There are a total of 3,421,083 rows of data and the missing values makes about 6% of the total rows of data. Since we are trying to use this analysis to investigate customer purchasing behavior, I am not sure how the rows with the missing value for the "days_since_prior_order" can add value to our analysis and I don't think removing these rows would damage the data; So I decided to removed these rows:


*** Correction: I undid the removal of the null values after the feedback from pedro:"These are all valid data points so removing them would be a mistake. Here I would leave it as null or use a number like -1 to indicate the first order and so the data is not mixed."

In [41]:
df_ords_firstorders_replaced = df_ords['days_since_prior_order'].fillna(value = -1, inplace = True)

In [61]:
# I noticed here that instead of having null values I have string NaN values that don't show up as null here:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_day_of_week         0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [62]:
# so to fix the NaN values, I replaced them with -1 using the .replace() function.
df_ords['days_since_prior_order'].replace({'Nan': -1})

0          -1.0
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64

In [63]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                    -1.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [64]:
# to check my work I counted the values in this column and I have the same number of -1s as I had nulls and NaNs! 
df_ords['days_since_prior_order'].value_counts()

 30.0    369323
 7.0     320608
 6.0     240013
 4.0     221696
 3.0     217005
 5.0     214503
-1.0     206209
 2.0     193206
 8.0     181717
 1.0     145247
 9.0     118188
 14.0    100230
 10.0     95186
 13.0     83214
 11.0     80970
 12.0     76146
 0.0      67755
 15.0     66579
 16.0     46941
 21.0     45470
 17.0     39245
 20.0     38527
 18.0     35881
 19.0     34384
 22.0     32012
 28.0     26777
 23.0     23885
 27.0     22013
 24.0     20712
 25.0     19234
 29.0     19191
 26.0     19016
Name: days_since_prior_order, dtype: int64

In [65]:
# CORRECTION: Leaving this here but I am not going to remove the null values in the clean dataframe. Replaced them with -1
# Removing the rows with missing values:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [53]:
df_ords_clean.shape

(3214874, 7)

In [54]:
# Question 8: Address the duplicates
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [56]:
df_ords_dups.shape

(0, 7)

There are no full duplicates found in this set, which is great. There might be rows that share the same order_id and user_id or even the department_id and this is normal since it can simply show that a customer purchase multiple items from the same department in the same order.

In [78]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                    -1.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [92]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [93]:
# df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [94]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                    -1.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [95]:
# Question 9: Export the final cleaned datafarmes: 
# CORRECTION: I am replacing the df_ords_clean with 

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'prods_clean.csv'))

df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean.csv'))

In [96]:
df_ords

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                    -1.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [89]:
df_test = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean.csv'), index_col = False)

In [97]:
df_test.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                    -1.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0